norse/torch/module/snn_cg.py

```diff
@@ -0,0 +1,81 @@
+import torch
+from typing import Any, Callable, List, Optional, Tuple
+
+from norse.torch.functional.lif import LIFFeedForwardState
+from norse.torch.module.snn import SNN, FeedforwardActivation
+
+# ===================================================================
+# Implement various SNN cells using CUDA Graphs
+# <https://pytorch.org/blog/accelerating-pytorch-with-cuda-graphs/>
+# ===================================================================
+
+class SNNCellCG(SNN):
+    """
+    Initializes a feedforward neuron cell *without* time.
+    Accelerated forward pass using Pytorch with CUDA Graphs.
+    <https://pytorch.org/blog/accelerating-pytorch-with-cuda-graphs/>
+
+    Parameters:
+        activation (FeedforwardActivation): The activation function accepting an input tensor, state
+            tensor, and parameters module, and returning a tuple of (output spikes, state).
+        state_fallback (Callable[[torch.Tensor], Any]): A function that can return a
+            default state with the correct dimensions, in case no state is provided in the
+            forward pass.
+        p (torch.nn.Module): The neuron parameters as a torch Module, which allows the module
+            to configure neuron parameters as optimizable.
+        dt (float): Time step to use in integration. Defaults to 0.001.
+        activation_sparse (Optional[FeedforwardActivation]): A Sparse activation function - if it exists
+            for the neuron model
+    """
+    def __init__(
+        self,
+        activation: FeedforwardActivation,
+        state_fallback: Callable[[torch.Tensor], torch.Tensor],
+        p: Any,
+        dt: float = 0.001,
+        activation_sparse: Optional[FeedforwardActivation] = None,
+    ):
+        super(SNNCellCG, self).__init__(
+            activation=activation,
+            state_fallback=state_fallback, 
+            p=p,
+            dt=dt, 
+            activation_sparse=activation_sparse
+        )
+        # configure internal CUDAGraph
+        self.config = True 
+
+        # graphed callable
+        self.activation_cg = None
+
+
+    def forward(self, input_tensor: torch.Tensor, state: Optional[Any] = None):
+        state = state if state is not None else self.state_fallback(input_tensor)
+
+        if self.config:
```

Comment:
Yes, this seems to be a good idea, but what happens if the user decides to give a pre-defined input state? I assume that in order to catch such cases. we would still need to check for the configured CUDA graph?

Reply:
Good point. But won't the user be able to use the `state_fallback` function to initialize the CUDA graph?

At the moment, this is not a pretty solution because the user would have to access the hidden `self.state_fallback` function. But I could imagine "raising" the `state_fallback` to a method in the SNN* classes (`def initialize_state(...)`?), so that the user will have a transparent way to perform the necessary initializations - if custom states are needed.

Reply:
Okay, I moved the CUDA Graph configuration into the `state_fallback` function. Furthermore, I allowed to pass an additional`LIFFeedForwardState` to the function, which enables the user to pre-define states and still configure the CUDA Graph. Note however, that the user needs to call `initial_state(data, pre_defined_state)` manually to set up the graph if she/he wants to use pre-defined states.

Reply:
Thanks!